File: Data Structures & Algorithms/permutations/submission-1.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> result;
		function<void(vector<int>)> backtrack;
		backtrack = [&](vector<int> curr) {
			if (curr.size() == nums.size()) {
				result.push_back(curr);
			}

			for (const auto& num : nums) {
				if (find(curr.begin(), curr.end(), num) == curr.end()) {
					curr.push_back(num);
					backtrack(curr);
					curr.pop_back();
				}
			}
		};
		backtrack(vector<int>{});
		return result;
    }
};

File: Data Structures & Algorithms/subsets/submission-1.go
func subsets(nums []int) [][]int {
    var result [][]int
    var curr []int
    var backtrack func(startIdx int)
    backtrack = func(startIdx int) {
        tmp := make([]int, len(curr))
        copy(tmp, curr)
        result = append(result, tmp)
        for i := startIdx; i < len(nums); i++ {
            curr = append(curr, nums[i])
            backtrack(i+1)
            curr = curr[:len(curr)-1]
        }
    }
    backtrack(0)
    return result
}
